scalableminds/webknossos
# Add Free Area selection to AI Quick Select (SAM squiggly lines)

This PR adds a new free area selection mode to the AI quick selection. It uses a free line / area draw tool as already used by the area measurement or trace selection tool. The selected area is then transformed into a mask and input into the SAM prediction. An example of how the mask should work can be found in this [notebook](https://github.com/facebookresearch/segment-anything/blob/main/notebooks/onnx_model_example.ipynb) under "Example mask input". 

It does not explicitly state what area the mask covers of the embedding put into the prediction. Still, I figured that the area covered by the mask is the area between the input coordinates (in the notebook `input_point` and in our code in `quick_select_ml_saga.ts` the coordinates are called `onnxCoord`). This seemed to work in the first free tries.

### URL of deployed dev instance (used for testing):
- https://samsquigglylines.webknossos.xyz/

### Steps to test:
- ~~I suggest testing locally as the SAM setup is easier this way:~~
  - ~~In the `application.conf` configure `segmentAnything` correctly to setup the SAM backend and set `segmentAnythingEnabled` to true.~~
- Feel free to test the quick select on the dev instance. SAM is enabled on these instances by default :tada: 
- Open any annotation
- Select the quick select tool
- A new button group should open up once the AI version is selected. Select the second "free area" version.
- Draw around like with the area measurement or trace tool. The tool should predict the encircled area and not a rectangular one.
- Play a little around with the tool and also test the rectangle version to confirm it is still working as expected.  

### TODOs:
- [x] Only show the tool switch buttons when the AI version is selected
- [x] Adjust icons of tool switch buttons
- [x] Keep the main tool button active when sub (are-) variation is selected.
    - [x] Also do this for the new measurement tool.  
- [ ] Test for edge cases
- [x] Add a better mouse pointer for the area select variation of the quick select

### Suggested Refactorings:
- [x] Rename `VolumeLayer` to `TemporaryVolumeAnnotation`.

### Issues:
- fixes [#7339](https://github.com/scalableminds/webknossos/issues/7339)

------
(Please delete unneeded items, merge only when none are left open)
